### DISCUSSION TITLE
Regex Manager depName mismatch when depName contains the value being updated

### DISCUSSION THREAD

**Discussion by jpredmond** [renovatebot/renovate#24533] [Request Help] [answered]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

36.14.0

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

I am trying to get this working for the first time

## Background:
Our GitLab CI configuration file contains a number of variables that specify container image URLs, so that those images can be referenced via the variables in the "image:" sections of various jobs that use the same image without duplication.  The standard managers do not interpret those variables, so we are attempting to use a regexManager to update these images.

**Given** a .gitlab-ci.yml file with the following variable declaration:
```yaml
variables:
  CUSTOM_IMAGE_RHEL7_X86_64: registry.gitlab.mycompany.com:443/our_group/a_subgroup/our_project/some_folder/image_name_ending_with_rhel7-x86_64:6
```

**And given** the following in our renovate.json:
```json
{
    "regexManagers": [
        {
            "description": [
                "All GitLab CI variables starting with 'CUSTOM_IMAGE_' referring to ",
                "images served from the our_project project container registry"
            ],
            "fileMatch": ["^\\.gitlab-ci\\.yml$"],
            "matchStrings": [
                "\\n\\s*CUSTOM_IMAGE_\\S+: (?<registryUrl>registry.gitlab.mycompany.com:443)/(?<depName>our_group/a_subgroup/our_project/some_folder/[^:]*?):(?<currentValue>\\d+)\\s"
            ],
            "datasourceTemplate": "docker"
        }
    ]
}
```

## Expected behavior:

We expect that the tag (":7") in the variable value should be changed to the latest available tag (i.e., ":8").

## Current behavior:

The update fails, with the following message in the debug log:
```
"manager":"regex","packageFile":".gitlab-ci.yml","currentDepName":"our_group/a_subgroup/our_project/some_folder/image_name_ending_with_rhel7-x86_64","newDepName":"our_group/a_subgroup/our_project/some_folder/image_name_ending_with_rhel8-x86_64","msg":"depName mismatch"
```

## Discussion:
So apparently some step in the process is applying the change from "7" to "8" to the depName instead of (or in addition to) applying it to the tag.

This behavior seems similar if not identical to the one documented in [Issue 8061](https://github.com/renovatebot/renovate/issues/8061), but that issue is marked as fixed.  I don't know if there could have been a regression that caused the issue to reappear, or if this is a different root cause with the same symptom.  In some discussions I found regarding that issue I saw a suggestion to add an autoReplaceStringTemplate setting to the regexManager, so I tried adding this line:
```json
   "autoReplaceStringTemplate": "{{{depName}}}:{{{newValue}}}",
```

This changed the error somewhat but didn't solve it.  The resulting "depName mismatch" error indicated that with this addition to the configuration, the depName was being changed to a different depName in the same file.
```
"manager":"regex","packageFile":".gitlab-ci.yml","currentDepName":"our_group/a_subgroup/our_project/some_folder/image_name_ending_with_rhel7-x86_64","newDepName":"our_group/a_subgroup/our_project/some_folder/image_name_ending_with_ubuntu_20_04-x86_64","msg":"depName mismatch"
```

In my [minimal reproduction repository](https://github.com/jpredmond/renovate-depName-mismatch-repro) I am demonstrating both the original problem and the unsuccessful workaround at the same time, using variables starting with "IMAGE_TEST_" and "WORKAROUND_TEST_" respectively.

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: depName mismatch (branch="renovate/jpredmond-type_1_image-2.x")
{
  "manager": "regex"
  "packageFile": ".gitlab-ci.yml"
  "currentDepName": "jpredmond/type_1_image"
  "newDepName": "jpredmond/type_2_image"
}

WARN: Error updating branch: update failure (branch="renovate/jpredmond-type_1_image-2.x")
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction in the description above

**Comment by rarkins** [comment]
Can you create a second reproduction repo with only _one_ failing dependency, and then also post the logs for _that_ instead? Logs with `redacted` are always harder to read, sometimes redact relevant info. You can install the reproduction into the hosted github.com app and get logs that way so nothing needs redacting.

**Reply by jpredmond** [reply]
OK, I created a truly minimal reproduction: https://github.com/jpredmond/renovate-depName-mismatch-Minimal-reproduction

Here is a link to the resulting failed job (I'm not sure if you can see it?) https://developer.mend.io/github/jpredmond/renovate-depName-mismatch-Minimal-reproduction/-/job/1d36cf0f-90cf-495e-84a6-103601a9daa3

I will update the logs section above with the error message from the job.  I don't know if it is sufficient - the debug logs I see in our self-hosted instance seem more detailed, but maybe it's just a matter of formatting?

**Comment by rarkins** [comment, marked as answer]
Converted to bug issue: https://github.com/renovatebot/renovate/issues/24559
